force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,11 +138,39 @@ public with sharing class DMLOperations {
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
-        // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+        
+        List<Account> accounts =[SELECT Id, Name                                    
+                                FROM Account 
+                                WHERE Name = :accountName];
+        Account acct = new Account();                        
+        if (!accounts.isempty()){                   //checks for existing account
+            acct = accounts[0];            
+        } else {                                    // otherwise, create a new Account
+            acct.Name = accountName;
+        }   
+        upsert acct;                        		// Upsert the Account
+                                
+        List<Opportunity> oppList = new List<Opportunity>([SELECT Id, Name FROM Opportunity WHERE Name IN :oppNames]);
+        Set<String>oppNamesSet = new Set<String> (oppNames);
+        if (!oppList.isempty()){
+            for(Opportunity opp : oppList){
+                opp.AccountId = acct.Id;
+                oppNamesSet.remove(opp.Name);
+            }
+        }
+        if (!oppNamesSet.isempty()){
+            for (String oppName : oppNamesSet){        // Loop through each Opportunity name
+                Opportunity opp = new Opportunity(
+                    Name = oppName,                 // Create an Opportunity with the given name
+                    StageName = 'Prospecting',
+                    CloseDate = (Date.today()).addDays(7),                   
+                    AccountId = acct.Id);    
+                oppList.add(opp);               // Add the Opportunity to the list                 
+            }
+        }
+    upsert oppList;
```

Comment:
I think the way you have done this here works by taking the list and removing the value you found. The only thing you might want to consider is where you would handle if you needed to make some updates to the "existing" opportunities that you queried, if you needed to. In this case they are not really changed at all so it does not really matter.  This way works and it is probably even less heap space used than my example!